### PR TITLE
Dynamic fields

### PR DESCRIPTION
**Pull request by Anbranin**
I made this PR as small as I could. 

As part of the review I really recommend trying the JS out yourself.

### CODE REVIEW

**Comment by jforgue** [comment]
Travis an clodecimate issues!

**Comment by Anbranin** [comment]
No shit

**Comment by jforgue** [comment]
Nit pick, but I think there should be should be space between the text and text boxes. Super nit pick, it would also look nice if the text boxes ended at the same point instead of being the same length.

Screen Shot 2019-03-27 at 11 52 30 AM

**Comment by Anbranin** [comment]
So, it did require me to make this PR _even larger_, but I've managed something. You're right, it didn't look great, because it didn't follow bootstrap's guidelines.

What `form-inline` is really designed for is a single-line form where you use your own padding elements, not really what we have here.. And I think the reason a normal `form-group` wasn't working is because text-align was set to `center` for _all_ text, so anything other than `form-inline` looked awful. I changed that within this little bit so that the regular ole bootstrap form-group would work, and I think it looks nice.

I would have definitely put this on a different PR, but it makes it only marginally bigger.

Screen Shot 2019-04-02 at 1 46 50 PM

**Comment by Anbranin** [comment]
aaaaaaaaaaah ⚡️

File: app/models/field.rb [hunk 1]
```diff
@@ -9,22 +9,21 @@ class Field < ApplicationRecord
...
+  validates :options, presence: true, if: -> { data_type == 'options' }
+  validates :data_type, inclusion: { in: DATA_TYPES }
+  validates :number, :prompt, presence: true
   # Equivalent to
   # validates :form, uniqueness: { scope: :number, allow_blank: true }
   # but the point is that the number is unique respective to the form,
   # not the other way around
   validates :number, uniqueness: { scope: :form,
-                                   if: -> { form.present? } }
+                                   if: -> { form_id.present? } }
```
Comment (dfaulken) [inline app/models/field.rb]: Curious why this change was needed.

File: app/models/form.rb [hunk 2]
```diff
@@ -37,6 +37,6 @@ def draft_belonging_to?(user)
   end
 
   def form_name
-    name.parameterize
+    name.try(:parameterize)
```
Comment (dfaulken) [inline app/models/form.rb]: Confused by this as name is validated for presence.

File: app/models/field.rb (hunk 1 above)
Comment (Anbranin) [reply app/models/field.rb]: I think I was trying to solve a problem, tried 3 things, one thing worked, left the other 2. I took it away and ran the tests and they all passed, so I'm guessing this wasn't actually necessary.

File: app/models/form.rb (hunk 2 above)
Comment (Anbranin) [reply app/models/form.rb]: Correct, and I think this may not be a problem in production given the html validations we have in place. However, this method is called on `new` objects as well as persisted objects--because of the way friendly-id works, it tries to create a slug immediately. This will probably never be an issue in production, but when I was playing in the console while developing I noticed this:
```
f = Form.new
=> #<Form id: nil, name: nil, intro_text: nil, created_at: nil, updated_at: nil, email: nil, reply_to: nil, slug: nil>
f.valid?
Traceback (most recent call last):
        2: from (irb):4
        1: from app/models/form.rb:40:in `form_name'
NoMethodError (undefined method `parameterize' for nil:NilClass)
```
Uncool, in my opinion, because it prevents the model validations from working as intended. Although you can certainly run all the current tests without that nil-handling, I believe it makes developing nicer and might prevent some weirdness later.

Comment (dfaulken) [reply app/models/form.rb]: OK, makes sense and agreed. My concern here was just that IIRC we use the parameterized name for the routing? So where I was going with this is that it could do weird things to the URL if having no name while you're making a draft for a new form is part of the new workflow.
As long as this is just to accommodate dev/test, then no worries.

Comment (Anbranin) [reply app/models/form.rb]: Yeah, it would definitely do weird things. However the form, like you said, can't be saved without a name regardless. 
This will not only accommodate dev/test, but allow model validations to work properly in the absence of html validations (an unlikely scenario) in production.

Comment (dfaulken) [reply app/models/form.rb]: OK, cool

**Comment by Anbranin** [comment]
⚡️ 
comaaaaaaaaaannnnnnn

**Comment by Anbranin** [comment]
plz for the love of God ⚡️

File: app/assets/javascripts/form_sorting.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    const newField = $('<textarea placeholder="' + placeholder + '" class="form-control" rows="4">');
+    newField.appendTo(container.find('.options'));
+  } else {
+      container.find('.options').children().remove();
+  }
+}
+function togglePlaceholder(dataType, container) {
+  if (takesPlaceholder(dataType) == true) {
+    if (container.find('.placeholder').children().length == 0) {
+      const newField = $('<input class="form-control" type="text">')
```
[the remark this answers is not in the document]
Comment (Anbranin) [reply app/assets/javascripts/form_sorting.js]: `text_field_tag` is a Rails method and will not work inside a javascript file, as far as I'm aware.

[the remark this answers is not in the document]
Comment (Chauce) [reply app/assets/javascripts/form_sorting.js]: Oops my bad -- I thought I'd done it a couple weeks ago but I just checked and I was mistaken.
